docs/manual/howto/widget.rst

```diff
@@ -0,0 +1,496 @@
+
+InLoopPollText
+--------------
+
+The ``base.InLoopPollText`` class builds on the ``base._TextBox`` by adding a timer to
+periodically refresh the displayed text.
+
+Widgets using this class should override the ``poll`` method to include a function that
+returns the required text.
+
+.. note::
+    This loop runs in the event loop so it is important that the poll method does not
+    call some blocking function. If this is required, widgets should inherit the
+    ``base.ThreadPoolText`` class (see below).
+
+ThreadPoolText
+--------------
+
+The ``base.ThreadPoolText`` class is very similar to the ``base.InLoopPollText`` class.
+The key difference is that the ``poll`` method is run asynchronously and triggers a
+callback once the function completes. This allows widgets to get text from
+long-running functions without blocking Qtile.
+
+Mixins
+======
+
+As well as inheriting from one of the base classes above, widgets can also inherit one
+or more mixins to provide some additional functionality to the widget.
+
+PaddingMixin
+------------
+
+This provides the ``padding(_x|_y|)`` attributes which can be used to change the appearance
+of the widget.
+
+If you use this mixin in your widget, you need to add the following line to your ``__init__``
+method:
+
+.. code:: python
+
+    self.add_defaults(base.PaddingMixin.defaults)
+
+MarginMixin
+-----------
+
+The ``MarginMixin`` is essentially effectively exactly the same as the ``PaddingMixin`` but,
+instead, it provides the ``margin(_x|_y|)`` attributes.
+
+As above, if you use this mixin in your widget, you need to add the following line to your
+``__init__`` method:
+
+.. code:: python
+
+    self.add_defaults(base.MarginMixin.defaults)
+
+Configuration
+=============
+
+Now you know which class to base your widget on, you need to know how the widget
+gets configured.
+
+Defining Parameters
+-------------------
+
+Each widget will likely have a number of parameters that users can change to
+customise the look and feel and/or behaviour of the widget for their own needs.
+
+The widget should therefore provide the default values of these parameters as a
+class attribute called ``defaults``. The format of this attribute is a list of
+tuples.
+
+.. code:: python
+
+    defaults = [
+        ("parameter_name",
+         default_parameter_value,
+         "Short text explaining what parameter does")
+    ]
+
+Users can override the default value when creating their ``config.py`` file.
+
+.. code:: python
+
+    MyCustomWidget(parameter_name=updated_value)
+
+Once the widget is initialised, these parameters are available at
+``self.parameter_name``.
+
+The __init__ method
+-------------------
+
+Parameters that should not be changed by users can be defined in the ``__init__``
+method.
+
+This method is run when the widgets are initially created. This happens before
+the ``qtile`` object is available. This section is therefore appropriate for
+hard coding values.
```

Comment:
I don't really understand what this last sentence means.

Reply:
This is in the context of where to define parameters/attributes but, you're right, it doesn't add anything so will remove it.